application/cmdbabstract.class.inc.php

```diff
@@ -3580,7 +3580,17 @@ public function DisplayStimulusForm(WebPage $oPage, $sStimulus, $aPrefillFormPar
 			$oPage->add_ready_script(InlineImage::EnableCKEditorImageUpload($this, $sTempId));
 		} else {
 			//we can directly apply the stimuli
-			$bApplyStimulus = $this->ApplyStimulus($sStimulus); // will write the object in the DB
+			try {
+				$bApplyStimulus = $this->ApplyStimulus($sStimulus); // will write the object in the DB
+			}
+			catch (Exception $e) {
+				if ($sOwnershipToken !== null) {
+					// Release the concurrent lock, if any
+					iTopOwnershipLock::ReleaseLock($sClass, $iKey, $sOwnershipToken);
+				}
+				throw $e;
+			}
+
```

Comment:
This duplicates the lock release logic, I would rather try something like the following. But it needs another look to it. (Mind that it factorize lines 3594-3603 as well but I can't modify them in the PR)

```suggestion
			try {
				$bApplyStimulus = $this->ApplyStimulus($sStimulus); // Will write the object in the DB
			}
			catch (Exception $oException) {
				// Catch any exception happening during the stimulus
				$bApplyStimulus = false;
				throw $oException;
			}
			finally {
				if (!$bApplyStimulus) {
					// Throw an application oriented exception if necessary
					throw new ApplicationException(Dict::S('UI:FailedToApplyStimuli'));
				} else {
					if ($sOwnershipToken !== null) {
						// Release the concurrent lock, if any
						iTopOwnershipLock::ReleaseLock($sClass, $iKey, $sOwnershipToken);
					}

				return true;
			}

```

Reply:
This code does not have the same behavior as the one I wrote....